File: bestia-zoneserver/src/main/java/net/bestia/zoneserver/chat/SubCommandModule.java
package net.bestia.zoneserver.chat;

import net.bestia.messages.MessageApi;
import net.bestia.model.domain.Account;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * The sub modules are used to implement chat command modules which are used by
 * the {@link MetaChatCommand} to bundle together various chat commands.
 * 
 * @author Thomas Felix
 *
 */
abstract class SubCommandModule extends BaseChatCommand {

	public SubCommandModule(MessageApi akkaApi) {
		super(akkaApi);
		// no op.
	}

	/**
	 * Returns a matcher regexp pattern to check the command against.
	 * @return
	 */
	protected abstract Pattern getMatcherPattern();
	
	protected abstract String getHelpText();
	
	protected abstract void executeCheckedCommand(Account account, String text, Matcher matcher);
	
	@Override
	public void executeCommand(Account account, String text) {
		
		// Check if user has rights.
		if(account.getUserLevel().compareTo(requiredUserLevel()) < 0) {
			return;
		}
		
		final Matcher matcher = getMatcherPattern().matcher(text);

		if(!matcher.matches()) {
			sendSystemMessage(account.getId(), getHelpText());
			return;
		}
		
		executeCheckedCommand(account, text, matcher);
	}
	
}

File: bestia-zoneserver/src/main/kotlin/net/bestia/zoneserver/chat/SubCommandModule.kt
package net.bestia.zoneserver.chat

import net.bestia.model.account.Account
import net.bestia.zoneserver.actor.MessageApi

import java.util.regex.Matcher
import java.util.regex.Pattern

/**
 * The sub modules are used to implement chat command modules which are used by
 * the [MetaChatCommand] to bundle together various chat commands.
 *
 * @author Thomas Felix
 */
internal abstract class SubCommandModule(messageApi: MessageApi) : BaseChatCommand(messageApi) {
  /**
   * Returns a matcher regexp pattern to check the command against.
   * @return
   */
  protected abstract val matcherPattern: Pattern

  abstract override val helpText: String

  protected abstract fun executeCheckedCommand(account: Account, text: String, matcher: Matcher)

  override fun executeCommand(account: Account, text: String) {
    if (account.userLevel < requiredUserLevel()) {
      return
    }

    val matcher = matcherPattern.matcher(text)
    if (!matcher.matches()) {
      sendSystemMessage(account.id, helpText)
      return
    }

    executeCheckedCommand(account, text, matcher)
  }
}
